### DISCUSSION TITLE
Global error-handling mechanism

### DISCUSSION THREAD

**Discussion by ffxsam** [withastro/roadmap#523] [Proposal]
### Body

# Summary

A way for Astro apps to globally catch uncaught errors and handle them in some specific way.

# Background & Motivation

There are always bound to be uncaught errors, and while some logging services like Sentry can catch and log them, I believe the error is still thrown (or perhaps someone isn't using Sentry/Rollbar/etc). It would be nice if Astro (like Nuxt) could have a special `error.astro` page that the user gets pushed to in case of an uncaught error.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Provide a user-friendly, out-of-the-box way for an app to respond to uncaught errors
- Improve Astro's DX even more 💪 

# Example

Nuxt handles this in [an interesting way](https://nuxt.com/docs/getting-started/error-handling#rendering-an-error-page), by having a special `error.vue` page that the user is bounced to. It also provides the error details as props so we can choose to display part of that to the user, log somewhere else, etc.

**Comment by OfekA-IAI** [comment]
Anyone has been able to do such a thing with the current Astro API?

We want to catch all unhandled exceptions, but are unable to. 
Any help would be much appreciated.

**Comment by ematipico** [comment]
I believe that's where the 500 page comes into play, right?

**Reply by OfekA-IAI** [reply]
This does seem to work, is there any documentation regarding this feature?

**Comment by yaronveg** [comment]
This could actually help a lot 😮

**Comment by tropperstyle** [comment]
I don't think it is currently possible to even log to an external service that an error occurred during HTTP streaming (which the node adapter does by default). I opened https://github.com/withastro/roadmap/discussions/616

**Comment by anovichenk0** [comment]
I do it like this. 



middleware/error.ts
```ts
import { renderToStaticMarkup } from 'react-dom/server'

// render component to html
const renderErrorPage = async (status: number) => {
    const ComponentHTML = await ErrorComponent({ status })
    const Page = ComponentHTML

    return renderToStaticMarkup(Page)
}

export const error: MiddlewareHandler = async (context, next) => {
    try {
        const response = await next()

        // catches the error knowingly returned via Response
        if (response.statusText !== 'OK') {
            return new Response(await renderErrorPage(response.status || 500), {
                headers,
            })
        }

        return response
    } catch (e) {
        // catches all unhandled errors
        return new Response(await renderErrorPage(500), { headers })
    }
}
 ```

page.astro
```ts
const data = await fetchContent(dynamic)

if (!data)
    return new Response('Page not found', {
        status: 404,
    })
```
Error.tsx
```tsx
import global from '@/global.css?inline'

const styles = `${global}`

type Props = {
    status?: string | number
}

const Error = ({ status }: Props) => {
    return (
        <>
            <div className="relative flex h-screen items-center justify-center">
                <h1 className="h2 text-center">
                    {status && `${status} `}
                    {status && getReasonPhrase(status)}
                </h1>
            </div>

            <style>{styles}</style>
        </>
    )
}
```

**Reply by PELMENAP** [reply]
yea it's usfl for me

**Comment by florian-lefebvre** [comment]
You can either use a middleware or the 500.astro page

**Reply by PierBover** [reply]
Is there any example on how to use a middleware to catch global errors?

I'm testing this but the error never reaches the `catch` statement.

```js
export function onRequest (context, next) {
	try {
		return next();
	} catch (error) {
		console.log(error);
	}
}
```

Edit:

It works when using `await`.

```js
export async function onRequest (context, next) {
	try {
		return await next();
	} catch (error) {
		if (error instanceof NotAuthenticatedError) {
			return new Response(null, {
				status: 302,
				headers: {
					'Location': '/login'
				}
			});
		} else {
			throw error;
		}
	}
}
```
